internal/controller/reinstall/reinstall.go

```diff
@@ -0,0 +1,378 @@
+			// Add hold annotation for manual intervention (CNF-15719)
+		}
+
+		changed := setReinstallStatusCondition(clusterInstance, *reinstallRequestProcessedCondition)
+		if changed || updateEndtime {
+			updateErr := patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch)
+			if updateErr != nil {
+				if err == nil {
+					err = updateErr
+				}
+			}
+		}
+	}()
+
+	tasks := []struct {
+		Description string
+		Run         func(context.Context, *zap.Logger, *v1alpha1.ClusterInstance) (reconcile.Result, error)
+	}{
+		{"Validating reinstall request", r.ensureValidReinstallRequest},
+		{"Deleting rendered manifests", r.ensureRenderedManifestsAreDeleted},
+	}
+	for _, task := range tasks {
+		log.Info("Executing task", zap.String("Task", task.Description))
+		result, err = task.Run(ctx, log, clusterInstance)
+		if !result.IsZero() {
+			log.Info("ClusterInstance to be requeued")
+			return
+		}
+		if err != nil {
+			log.Error("Task failed", zap.String("Task", task.Description), zap.Error(err))
+			reinstallRequestProcessedCondition = reinstallRequestProcessedConditionStatus(
+				metav1.ConditionFalse, v1alpha1.Failed,
+				fmt.Sprintf("Encountered error executing task: %s. Error: %v", task.Description, err))
+			updateEndtime = true
+			return
+		}
+		log.Info("Task completed", zap.String("Task", task.Description))
+	}
+
+	log.Info("Finalizing reinstall request")
+	if err = r.finalizeReinstallRequest(ctx, log, clusterInstance); err != nil {
+		return
+	}
+	reinstallRequestProcessedCondition = reinstallRequestProcessedConditionStatus(
+		metav1.ConditionTrue,
+		v1alpha1.Completed,
+		"The reinstall process completed successfully and the cluster is ready to be reprovisioned.")
+
+	return
+}
+
+func isRequestProcessed(clusterInstance *v1alpha1.ClusterInstance) bool {
+
+	reinstallStatus := clusterInstance.Status.Reinstall
+	if reinstallStatus == nil {
+		return false
+	}
+
+	cond := meta.FindStatusCondition(reinstallStatus.Conditions, string(v1alpha1.ReinstallRequestProcessed))
+	if cond == nil {
+		return false
+	}
+
+	if reinstallStatus.ObservedGeneration == clusterInstance.Spec.Reinstall.Generation {
+		return true
+	}
+
+	isProcessed := cond.Reason == string(v1alpha1.Completed) ||
+		cond.Reason == string(v1alpha1.Failed) ||
+		cond.Reason == string(v1alpha1.TimedOut)
+
+	return reinstallStatus.InProgressGeneration == clusterInstance.Spec.Reinstall.Generation &&
+		!reinstallStatus.RequestEndTime.IsZero() &&
+		isProcessed && cond.Status != metav1.ConditionUnknown
+}
+
+func isNewRequest(clusterInstance *v1alpha1.ClusterInstance) bool {
+
+	reinstallStatus := clusterInstance.Status.Reinstall
+
+	if reinstallStatus == nil || reinstallStatus.InProgressGeneration != clusterInstance.Spec.Reinstall.Generation {
+		return true
+	}
+
+	return false
+}
+
+// ensureValidReinstallRequest checks if cluster reinstall request is valid
+func (r *ReinstallHandler) ensureValidReinstallRequest(
```

Comment:
Is there anything else that needs to be validated? Shouldn't we check if the instance has been installed successfully?

Reply:
Good suggestion about checking if the instance is installed - I think I will add this validation step in a separate PR (to allow for faster dev tests). I'll create a Jira for this.